Use colony for map and vehicle item storage

ifreund opened the pull request:
#### Summary
SUMMARY: Infrastructure "Use colony for map and vehicle item storage."

#### Purpose of change
As described in #31338, colony has several advantages over the current `std::list`, the most important being that iterators are not invalidated by erasure.  This will make the item management code much more stable and has already cleaned up one nasty hack with processing activity items.

#### Describe the solution
- [x] Replace the `std::list<item>` stored in submaps and vehicle parts with a `colony<item>`.
- [x] Tweak the api of`item_stack` to reflect this change.
- [x] Make necessary changes to the rest of the code to reflect this
- [x] Delete `item_stack::operator[]` and refactor
- [x] Delete map and vehicle item access/removal functions that take an index and refactor
- [x] Comb through code to make sure iterators are used wherever possible
- [x] Rewrite `invlet_test.cpp`

#### Additional context
I've done my best to Include What I Use, but I've certainly missed some includes.  Maybe I will try and get the IWYU tooling set up when this is done.

jbytheway commented:
> colony has several advantages over the current std::list, the most important being that iterators are not invalidated by erasure

That's not true.  `list` iterator invalidation is basically exactly the same as for `colony`.  The advantages of `colony` are better memory layout and fast-ish random access.

ifreund commented on src/submap.h:
```diff
@@ -288,8 +289,9 @@ struct maptile {
+        // Assumes there is at least one item
         const item &get_uppermost_item() const {
-            return sm->itm[x][y].back();
+            return *std::prev( sm->itm[x][y].cend() );
```
Yeah that's exactly what it's for.

And yeah that's for a future PR.  I've also considered rewriting some of the inventory code to display alphabetically since storage on the map is no longer ordered.

ifreund commented on src/map.h:
```diff
@@ -1579,10 +1577,8 @@ class map
-                                   std::list<item>::iterator end );
-        std::unique_ptr<vehicle> add_vehicle_to_map( std::unique_ptr<vehicle> veh,
-                bool merge_wrecks );
+        void add_light_from_items( const tripoint &p, map_stack::iterator begin, map_stack::iterator end );
```
Yeah you're totally right, that was just a typo after replacing too many iterators.

ifreund commented on src/map.cpp:
```diff
@@ -4483,49 +4471,35 @@ void map::update_lum( item_location &loc, bool add )
+    if( temp.process( nullptr, location, activate, insulation, flag ) ) {
+        // Item is to be destroyed so erase the null item in the map stack
+        // unless it was already destroyed by processing.
+        if( std::find_if( items.begin(), items.end(), [target]( const item & it ) {
```
Yep! I've already got the colony `get_X_from_X` functions locally, will be doing more refactoring with them once indexes are finally obsoleted.

ifreund commented on src/json.h:
```diff
@@ -564,6 +591,13 @@ class JsonOut
             write_as_array( container );
         }
 
+        // special case for colony as it uses `insert()` instead of `push_back()`
```
Yeah that doesn't really make sense.  Will fix.

ifreund commented:
>That's not true. list iterator invalidation is basically exactly the same as for colony. The advantages of colony are better memory layout and fast-ish random access.

Um, yeah may not have totally thought that one through.  The meat of this PR is turning out to be not swapping in the colony but getting rid of all indexes to items on the map and vehicles.

ifreund commented:
>Have you tackled/tested activity serialization yet? Didn't see anything related to that as I perused.

Haven't yet and it works fine for now since `item_location` serialization still works.  I will be tweaking that some before this PR is finished to use the colony `get_X_from_X` functions.

jbytheway commented:
Oh, one more thing: if you can, it may be worth checking memory consumption impact of these changes.  An `item` is quite big`, so a `colony<item>` with only one `item` in it might use quite a lot of memory depending on how much space it allows for expansion.  If very small stacks of items are common this might be an issue.

jbytheway commented on src/activity_handlers.cpp:
```diff
@@ -517,12 +495,16 @@ static void set_up_butchery( player_activity &act, player &u, butcher_type actio
+
+    // We have a valid target, so preform the full finish function
+    // instead of just selecting the next valid target
+    act.index = false;
```
IS `act.index` a `bool`?  It seems unlikely...

jbytheway replied:
Ah, I see later that you are treating it as a `bool`.  I can't decide whether it's clearer to assign `0` or `false` here. Hmm.

ifreund replied:
Yeah I agree that it's not super clean.  I considered adding either a `bool` or an `std::vector<bool>` to `player_activity` but decided that was outside the scope of this PR

jbytheway replied:
I have an aspiration to make accessing things inside `player_activity` cleaner in general, with some vague ideas of how to go about it.  Looking at your code here makes me want to get to that, but I probably won't soon.

jbytheway commented on src/activity_handlers.cpp:
```diff
@@ -1022,6 +1005,21 @@ static void butchery_quarter( item *corpse_item, player &p )
+    item_location &target = act->targets.back();
+
+    // Corpses can disappear (rezzing!), so check for that
+    if( !target || !target->is_corpse() ) {
```
The `operator bool` on `item_location` is definitely doing the wrong thing in general.  That's not really your fault, but this can totally end up pointing at the wrong item (I think there have been bug reports about butchery doing that already).  I think the only way to fix this sensibly is with the `weak_ptr` idea Kevin has discussed.  But that's orthogonal to your changes.

ifreund replied:
I think what it's doing is actually OK here for now.  It just searches at its stored location for an item matching its stored pointer, returning true if such an item is found.

I agree the`weak_ptr` plan is definitely a lot cleaner and more robust and may look at implementing that once this get wrapped up.

jbytheway replied:
Bear in mind that `colony` deliberately tries to reuse memory, so there is a good chance that there will be some different `item` at the same address if one is removed and another added, and it has no way to tell the difference.

ifreund replied:
Right, but I don't think there's anything we can do about it yet without the `weak_ptr` trick aside from hope it's not a corpse.

ifreund commented:
the invlet tests are currently broken, will look into a fix.

ifreund commented:
This PR is virtually complete now.  The only remaining major task is a rewrite of `invlet_test.cpp` to account for the new indexless, unordered nature of items on the map/in vehicles.

jbytheway commented on src/item_stack.h:
```diff
@@ -34,7 +34,8 @@ class item_stack
         virtual iterator erase( const_iterator it ) = 0;
         virtual void clear();
-        item &front();
+        // Will cause a debugmsg if there is more than one item at the location
```
"more than one" -> "not exactly one"

anothersimulacrum commented on src/faction_camp.cpp:
```diff
@@ -1909,25 +1909,29 @@ static std::pair<size_t, std::string> farm_action( const tripoint &omt_tgt, farm
-                    const item &seed = farm_map.i_at( pos ).front();
-                    if( farm_valid_seed( seed ) ) {
+                if( farm_map.furn( pos ) == f_plant_harvest ) {
+                    // Can't use item_stack::only_item() since there might be fertilizer
```
It should be checking for the `GROWTH_HARVEST` flag, but it looks as though this whole thing is largely a duplicate of `iexamine::harvest`, and is out of date with the current state of what `iexamine::harvest` does.

ifreund replied:
This function led me to believe that fertilizer stays on the tile.  It could also easily be outdated though.
https://github.com/CleverRaven/Cataclysm-DDA/blob/4d34142d4cd0666b48bfd8ffce67a22af63608da/src/iexamine.cpp#L2066-L2082

jbytheway replied:
You're right.  I did read `iexamine::fertilize_plant` but I missed that line.

ifreund commented on src/game.cpp:
```diff
@@ -5119,23 +5119,22 @@ bool game::forced_door_closing( const tripoint &p, const ter_id &door_type, int
-                m.i_rem( x, y, 0 );
+        for( item &it : m.i_at( x, y ) ) {
+            if( it.made_of( LIQUID ) ) {
+                m.i_rem( x, y, &it );
```
Thanks, I should really know better

ifreund commented:
Alright, I think this is ready to go.

The new invlet test uses a hack to enable item UUIDs, but It seems to actually be more robust than what we had previously as the old test was giving false positives after #30603.

Sorry that this is so large.  If needed there are some parts that I could probably break off, but I wont have time to do so for a couple days.

jbytheway commented on tests/invlet_test.cpp:
```diff
@@ -46,6 +46,32 @@ enum test_action {
 
+// This is a massive hack that makes this test work without totally rewriting it after #31406
+// The number of null items in item.components is used as a UUID for the purposes of this test
+static void set_uuid( item &it, int uuid )
```
`uuid` implies a specific type of 128-bit id.  Maybe just call it an `id` instead.

ifreund replied:
will do

jbytheway commented on tests/invlet_test.cpp:
```diff
@@ -46,6 +46,32 @@ enum test_action {
+// The number of null items in item.components is used as a UUID for the purposes of this test
+static void set_uuid( item &it, int uuid )
+{
+    it.components = std::list<item>( static_cast<size_t>( uuid ), item() );
```
I believe `item` objects have support for arbitrary properties, which is probably a neater solution than using `components`.  However, I'm not sure how it works.  There's `has_property` and `get_property` functions, but I don't see how you set properties...

ifreund replied:
Properties won't work as they are stored on the itype level not per item instance.  I chose `components` because it's the only thing I could find that wouldn't prevent items with different ids from stacking, which is important for part of this test.

jbytheway replied:
Ah yes, makes sense.

ifreund commented:
Added a comment explaining why the unique id hack was implemented the way it is using `components` and rebased to clean up all the commit messages before merge.

ifreund commented:
Some rough numbers for the performance gains of `map::process_active_items()` with this PR applied:

Low number of active items:
Pre colony: 5.5%
Post colony: 3.2%

Lots of active items:
Pre colony: 28.7%
Post colony: 18.0%

The next target for optimizing item processing is `item::has_flag()` which could be made much faster by replacing the current `std::set<std::string>` with our enum bitset introduced in #29368

ifreund commented:
Profiling results part 2, Memory usage:

A worst case scenario for colony is 1 item per tile as the minimum group size is 8.

I tested this case by spawning and killing 2600 moose, to create a one corpse per tile test case.

![Screen Shot 2019-06-18 at 5 05 30 PM](https://user-images.githubusercontent.com/12723818/59696755-e31abb00-91ec-11e9-94ed-23a3a54865ba.png)

Memory usage of `submap::load()`is a little over 8 times higher after applying this PR.
Pre colony: 845.42KB, 0.4% of total memory usage.
Post colony: 8.06MB, 3.5% of total memory usage.

Why is it not exactly 8 times higher?

I haven't looked into the  gritty details, but colony has more memory overhead than just the stored types to increase performance. For example, there is a list to keep track of which elements have been erased so those memory locations can be reused for future insertion.

What can we do to mitigate this?

The simplest option is setting the starting group size to something less than 8 (minimum of 3).  I haven't looked into how this would effect performance yet, but will do some investigation.

Another option would be to store one `colony<item>` per submap and a `colony<item *>` for every maptile.  This would be more optimal for memory usage, but increases code complexity and may affect performance.

Do we need to do something about this?

I'm not sure, the memory usage isn't crazy and the current setup is probably better for performance than optimizing for lower memory usage.  Up to @kevingranade though really.  Regardless, I'd like to address that issue (if needed) in a follow up PR assuming that this is acceptable to merge as is.

jbytheway commented:
Thanks for testing that.  Another option is to look into making the `item` objects smaller.  The way they are currently structured is quite wasteful in most cases.

ifreund commented:
Conflicts resolved

ZhilkinSerg commented:
> 
> 
> Conflicts resolved

You wish :)

ifreund commented:
>You wish :)

Look again :)

ZhilkinSerg commented on src/iexamine.cpp:
```diff
@@ -2979,23 +2996,14 @@ void iexamine::tree_maple_tapped( player &p, const tripoint &examp )
-            }
-
+        case HARVEST_SAP: {
+            liquid_handler::handle_liquid_from_container( *container, PICKUP_RANGE );
```
Looks like `*container` can be a `nullptr` and there is no check for that.

ifreund replied:
This option is only available if `has_sap` is true, and `has_sap` can only be true if a container was found.

ifreund replied:
I'm not saying this code isn't fragile and prone to break at some point, but it is currently correct I believe.
